cliquet/authorization.py

```diff
@@ -45,6 +45,13 @@ def permits(self, context, principals, permission):
                 principals,
                 get_bound_permissions=self.get_bound_permissions)
 
+        is_list_operation = (context.on_collection and
+                             'create' not in permission)
+        if not allowed and is_list_operation:
+            # Guest trying to reach a collection.
+            setattr(context, 'guest_principals', principals)
```

Comment:
why not initializing `context.guest_principals` with an empty list instead of adding an attribute on the fly ? doing a `setattr` is making your life harder to read that list for no obvious reason I can see